cli/src/ui/commands/Auth.tsx

```diff
@@ -0,0 +1,273 @@
+        method: "POST",
+        headers: {
+          "Content-Type": "application/x-www-form-urlencoded",
+        },
+        body: new URLSearchParams({
+          client_id: clientId,
+          audience,
+          scope,
+        }),
+      });
+
+      if (!response.ok) {
+        const errorData = await response.text();
+        throw new Error(`Failed to start device flow: ${errorData}`);
+      }
+
+      const data = (await response.json()) as DeviceCodeResponse;
+      setDeviceCode(data);
+      setIsLoading(false);
+
+      // Open the verification URI in the user's browser
+      await open(data.verification_uri_complete);
+
+      // Start polling for the token
+      startPolling(data);
+    } catch (err) {
+      setIsLoading(false);
+      setError(err instanceof Error ? err.message : String(err));
+    }
+  };
+
+  const startPolling = (deviceCodeData: DeviceCodeResponse) => {
+    setIsPolling(true);
+
+    const pollInterval = deviceCodeData.interval;
+    const expiresIn = deviceCodeData.expires_in;
+    const maxAttempts = Math.floor(expiresIn / pollInterval);
+
+    let attempts = 0;
+
+    const pollForToken = async () => {
+      if (attempts >= maxAttempts) {
+        setIsPolling(false);
+        setError("Authentication timed out. Please try again.");
+        return;
+      }
+
+      try {
+        const response = await fetch(`https://${auth0Domain}/oauth/token`, {
+          method: "POST",
+          headers: {
+            "Content-Type": "application/x-www-form-urlencoded",
+          },
+          body: new URLSearchParams({
+            grant_type: "urn:ietf:params:oauth:grant-type:device_code",
+            device_code: deviceCodeData.device_code,
+            client_id: clientId,
+          }),
+        });
+
+        const data = (await response.json()) as
+          | TokenResponse
+          | TokenErrorResponse;
+
+        if ("error" in data) {
+          if (data.error === "authorization_pending") {
+            attempts++;
+            setTimeout(pollForToken, pollInterval * 1000);
+          } else if (data.error === "slow_down") {
+            attempts++;
+            setTimeout(pollForToken, (pollInterval + 5) * 1000);
+          } else {
+            // An actual error occurred
+            setIsPolling(false);
+            setError(
+              `Authentication error: ${data.error_description || data.error}`
+            );
+          }
+        } else {
+          // Store tokens in secure storage
+          await TokenStorage.saveTokens(data.access_token, data.refresh_token);
+
+          resetDustClient();
+
+          setIsPolling(false);
+
+          setShowWorkspaceSelector(true);
+        }
+      } catch (err) {
+        setIsPolling(false);
+        setError(err instanceof Error ? err.message : String(err));
+      }
+    };
+
+    setTimeout(pollForToken, pollInterval * 1000);
+  };
+
+  const handleWorkspaceSelectionComplete = async () => {
+    try {
```

Comment:
Technically this try/catch is prohibited :-)

Reply:
removed